### PR TITLE
Several ADT improvements

### LINKED ISSUE #589: Unexpected behaviour with `match` statements
The following code snippet is accepted, even though it should be rejected:
```go
type list adt {
	Empty{}

	Cons{
		head any
		tail list
	}
}

ghost
decreases l
func length(l list) int {
	match l {
		case Empty{}:
			assert false // <--- assert false here should be reachable
			return 0
		default: 
			return 0
	}
}
```

### LINKED ISSUE #641: Upcasting of ADT instance is required before calling a method
In the book I am translating, I encountered the following problem.

We need to upcast "Suc{x}" to "Unary(Suc{x})" in order to call the Add function: "Unary(Suc{x}).Add(y)".
Here is the declaration of the Add function: "pure func (x Unary) Add(y Unary) Unary".

This is also the case with other functions and ADTs.

### LINKED ISSUE #685: ADT constructors do not apply implicit conversions
The following snippet is causes a verification error due to an inconsistent Viper program, but should verify successfully:

```
type X adt {
  C{ c any }
}

func foo() {
  x := C{5}
}
```

### LINKED ISSUE #686: Usability of ADTs
Currently, ADTs can be very cumbersome to use. From my experience in SCION, this mostly comes from 3 points:
- constructors of different ADT types must have different names, and a constructor cannot have the same name as a type. Here, it could be useful to allow the user to qualify a constructor with its type to resolve any ambiguities when present
- match expressions must have strongly pure sub expressions: this hurts the usability of match statements, specially when defining predicates
- ADT constructors must have named parameters

### PR DESCRIPTION
**Pull request by Felalolf**
This PR adds several requested improvements for adts and fixes some past issues with adts:

- the result of an adt constructor is now the declared type of the adt type definition (was a special clause type before)
- arguments to adt literals now consider implicit conversions
- match expressions can occur in assertions
- the names of adt clause fields can be omitted
- duplicate field names now cause a type error instead of a crash
- duplicate clause names now cause a type error instead of a crash
- adt clause name can be qualfied over the adt type 
- adt clause names do not cause name collision with other package entities (their precedence is set to lowest)
- set and sequence operations (except conversions) are now more forgiving
- importing adts is a bit more robust
- fixed an issue with reorderings of match statement clauses

Adt clauses imported from another package can now be unreachable. Currently, qualified names of types must have at most one dot. However, with the new design, accessing an adt clause may require 2 dots, e.g. `pkg.adtType.adtClause`. This should be fixed in a future PR.

Fixes #686 
Fixes #685 
Fixes #641 
Fixes #589

### CODE REVIEW

**Review by jcp19 — APPROVED** [review]
Patina! I just have a few minor comments

File: src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/GhostIdTyping.scala [hunk 1]
```diff
@@ -30,7 +30,7 @@ trait GhostIdTyping { this: TypeInfoImpl =>
     })
     case c: AdtDestructor => wellDefAndType(c.decl.typ)
     case _: AdtDiscriminator => LocalMessages(noMessages)
-    case _: MatchVariable => LocalMessages(noMessages)
+    case _: MatchVariable => LocalMessages(noMessages) // TODO
```
[the remark this answers is not in the document]
Comment (Felalolf) [reply src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/GhostIdTyping.scala:33]: yes, forgot about it. I will try to remember, probably some error messages

File: src/test/resources/regressions/issues/000686-6.gobra [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Any copyright is dedicated to the Public Domain.
+// http://creativecommons.org/publicdomain/zero/1.0/
+
+package pkg
+
+//:: ExpectedOutput(type_error)
+type X adt {
+  A { x int }
+  B { x int }
+}
```
[the remark this answers is not in the document]
Comment (Felalolf) [reply src/test/resources/regressions/issues/000686-6.gobra:10]: I am not convinced that this is a good idea, but we can discuss this at some point.

File: src/test/resources/regressions/issues/000655.gobra [hunk 3]
```diff
@@ -13,4 +13,24 @@ pure func foo(x pkg.List) bool {
...
+pure func foo3(x pkg.Tree) bool {
+  return match x {
+    case pkg.Tree.Leaf{}: true
+    case pkg.Branch{?left, pkg.Tree.Leaf{}}: true
+    case pkg.Tree.Branch{pkg.Tree.Leaf{}, ?right}: true
+    case _: false
+  }
+}
+
+*/
```
[the remark this answers is not in the document]
Comment (jcp19) [reply src/test/resources/regressions/issues/000655.gobra]: Nvm, brain fart. I realize now that it uses paths of length 3

**Pull request merged**